### PR TITLE
`loadwallet` endpoint

### PR DESCRIPTION
**Pull request by rorp**
Introduces `loadwallet`, and `listwallets` endpoints

### CODE REVIEW

File: app/server/src/main/scala/org/bitcoins/server/WalletRoutes.scala [hunk 1]
```diff
@@ -978,6 +979,46 @@ case class WalletRoutes(wallet: AnyDLCHDWalletApi)(implicit
...
+              .split(WalletStorage.ENCRYPTED_SEED_FILE_NAME)
+              .headOption
+              .flatMap(wn =>
+                if (wn.endsWith("-")) Some(wn.substring(0, wn.length - 1))
+                else None)
+          }
+        Server.httpSuccess(list)
+      }
+
+    case ServerCommand("getwalletname", _) =>
```
Comment (user411) [inline app/server/src/main/scala/org/bitcoins/server/WalletRoutes.scala]: `walletinfo` endpoint already does this

**Review by Christewart — CHANGES_REQUESTED** [review]
`loadwallet`/`listwallet` need some unit tests

**Comment by user411** [comment]
I imported a new wallet and the response is `{result: null, error: null}`. I think we should return the new wallet name as a success signal. That will match `loadwallet` returning wallet name.

**Comment by Christewart** [comment]
needs rebase

**Comment by Christewart** [comment]
needs rebase

**Comment by Christewart** [comment]
Accidentally closed, and needs rebase

File: app/server/src/main/scala/org/bitcoins/server/WalletHolder.scala [hunk 2]
```diff
@@ -0,0 +1,943 @@
...
+import java.time.Instant
+import scala.concurrent.{ExecutionContext, Future}
+
+class WalletNotInitialized extends Exception("The wallet is not initialized")
+
+class WalletHolder(implicit ec: ExecutionContext)
+    extends AnyDLCHDWalletApi
+    with Logging {
+
+  @volatile private var wallet: AnyDLCHDWalletApi = _
```
Comment (Christewart) [inline app/server/src/main/scala/org/bitcoins/server/WalletHolder.scala]: Don't use `null`

File: app/server/src/main/scala/org/bitcoins/server/WalletHolder.scala [hunk 3]
```diff
@@ -0,0 +1,943 @@
...
+
+  def isInitialized: Boolean = synchronized {
+    wallet != null
+  }
+
+  def replaceWallet(newWallet: AnyDLCHDWalletApi): Future[AnyDLCHDWalletApi] =
+    synchronized {
+      val oldWallet = wallet
+      wallet = null
```
Comment (Christewart) [inline app/server/src/main/scala/org/bitcoins/server/WalletHolder.scala]: Use `Option`

File: app/server/src/main/scala/org/bitcoins/server/util/CallbackUtil.scala [hunk 4]
```diff
@@ -1,47 +1,54 @@
...
+        .recover { case _: WalletNotInitialized => () }
     }
     lazy val onBlock: OnBlockReceived = { block =>
       logger.debug(
         s"Executing onBlock callback=${block.blockHeader.hashBE.hex}")
-      wallet.processBlock(block).map(_ => ())
+      wallet
+        .processBlock(block)
+        .map(_ => ())
+        .recover { case _: WalletNotInitialized => () }
```
Comment (Christewart) [inline app/server/src/main/scala/org/bitcoins/server/util/CallbackUtil.scala]: Should we log an error or warning message in these cases?

File: wallet/src/main/scala/org/bitcoins/wallet/Wallet.scala [hunk 5]
```diff
@@ -916,6 +915,29 @@ abstract class Wallet
...
+  }
+
+  override def getInfo(): Future[WalletInfo] = {
+    for {
+      accountDb <- getDefaultAccount()
+      walletState <- getSyncState()
+      rescan <- isRescanning()
+    } yield {
+      WalletInfo(
+        walletName = walletConfig.walletNameOpt.getOrElse(""),
```
Comment (Christewart) [inline wallet/src/main/scala/org/bitcoins/wallet/Wallet.scala]: nit: Maybe make a constant somewhere called 

```scala
val DEFAULT_WALLET_NAME: String = ""
```

perhaps put it in `WalletAppConfig`?

**Comment by Christewart** [comment]
`loadwallet` is not implemented in the `cli` project

**Comment by Christewart** [comment]
I ran into an error this morning when trying to switch back to the default wallet on #4500 . The error message

```
Wallet `` does not exist
```

```
[info] 2022-07-12T15:49:12UTC INFO [BitcoinSServerMain] Loading wallet
[info] 2022-07-12T15:49:12UTC INFO [Server] Unhandled error in server:
[info] java.lang.RuntimeException: Wallet `` does not exist
[info]  at org.bitcoins.server.BitcoinSServerMain.$anonfun$updateWalletConfigs$1(BitcoinSServerMain.scala:662)
[info]  at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:467)
[info]  at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
[info]  at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
[info]  at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[info]  at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
[info]  at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
[info]  at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
[info]  at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
[info]  at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
[info]  at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
[info]  at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
[info]  at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
[info]  at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
```

Screenshot from 2022-07-12 10-53-20

**Comment by rorp** [comment]
> I ran into an error this morning when trying to switch back to the default wallet on #4500 . The error message
> 
> ```
> Wallet `` does not exist
> ```

The CLI implementation is incorrect in #4500, use this PR for CLI

**Comment by user411** [comment]
When importing a new wallet, the `height` is initially set to zero. After doing a `loadwallet` and restarting the backend with that new wallet, the `height` is set to current blockHeight.

Still need the `import` endpoints to return the wallet name for a successful import.

**Comment by user411** [comment]
When I tried a `rescan` on an imported wallet without setting `ignoreCreationTime`, I got the following:
```
[info] 2022-07-14T15:13:08UTC INFO [DLCWallet$DLCWalletImpl] Starting rescanning the wallet from None to None useCreationTime=true
[info] 2022-07-14T15:13:29UTC ERROR [DLCWallet$DLCWalletImpl] Failed to rescan wallet
[info] java.lang.RuntimeException: No block headers in database.
[info] 	at org.bitcoins.chain.models.BlockHeaderDAO.$anonfun$findClosestToTime$5(BlockHeaderDAO.scala:238)
```

**Comment by user411** [comment]
May want to have `walletinfo` take an optional parameter of wallet name so details can be pulled on all wallets without loading them.

**Comment by user411** [comment]
Calling `loadwallet` during a wallet rescan causes a hard crash (full stop) on my local backend. When I restarted the backend, it loaded the wallet I last requested to `loadwallet`. When I `loadwallet`ed the scanning wallet, it did not resume rescanning the wallet. However, the `rescan` flag is still `true` on the wallet. Second restart of the backend I noticed
```
[info] 2022-07-15T14:47:03UTC INFO [DLCWallet$DLCWalletImpl] Starting rescanning the wallet from None to None useCreationTime=true
[info] 2022-07-15T14:47:23UTC ERROR [DLCWallet$DLCWalletImpl] Failed to rescan wallet
[info] java.lang.RuntimeException: No block headers in database.
```
and the backend did not load the wallet. Third restart on the backend cleared the `rescan` flag on the wallet and started ok.

**Comment by Christewart** [comment]
It seems when switching to a wallet that was rescanning last time we started it doesn't re-start the rescan.

Screenshot from 2022-07-18 08-10-22

**Comment by user411** [comment]
First `Launch Wallet`
```
[info] 2022-07-18T21:45:28UTC INFO [BitcoinSServerMain] Starting neutrino node sync, it took=565ms
[error] org.bitcoins.wallet.WalletNotInitialized: The wallet is not initialized
[info] 2022-07-18T21:45:28UTC ERROR [BitcoinSServerMain] Failed to startup server!
[info] org.bitcoins.wallet.WalletNotInitialized: The wallet is not initialized
```
Second `Launch Wallet`
```
[info] 2022-07-18T21:46:13UTC ERROR [BitcoinSServerMain] Failed to startup server!
[info] akka.stream.impl.io.ConnectionSourceStage$$anon$1$$anon$2: Bind failed because of java.net.BindException: [/127.0.0.1:9999] Address already in use
[info] Caused by: java.net.BindException: [/127.0.0.1:9999] Address already in use
[error] akka.stream.impl.io.ConnectionSourceStage$$anon$1$$anon$2: Bind failed because of java.net.BindException: [/127.0.0.1:9999] Address already in use
[error] Caused by: java.net.BindException: [/127.0.0.1:9999] Address already in use
```
Third, after restarting everything, was same as the first. I had a non-default wallet set.

**Comment by Christewart** [comment]
> First `Launch Wallet`
> 
> ```
> [info] 2022-07-18T21:45:28UTC INFO [BitcoinSServerMain] Starting neutrino node sync, it took=565ms
> [error] org.bitcoins.wallet.WalletNotInitialized: The wallet is not initialized
> [info] 2022-07-18T21:45:28UTC ERROR [BitcoinSServerMain] Failed to startup server!
> [info] org.bitcoins.wallet.WalletNotInitialized: The wallet is not initialized
> ```
> 
> Second `Launch Wallet`
> 
> ```
> [info] 2022-07-18T21:46:13UTC ERROR [BitcoinSServerMain] Failed to startup server!
> [info] akka.stream.impl.io.ConnectionSourceStage$$anon$1$$anon$2: Bind failed because of java.net.BindException: [/127.0.0.1:9999] Address already in use
> [info] Caused by: java.net.BindException: [/127.0.0.1:9999] Address already in use
> [error] akka.stream.impl.io.ConnectionSourceStage$$anon$1$$anon$2: Bind failed because of java.net.BindException: [/127.0.0.1:9999] Address already in use
> [error] Caused by: java.net.BindException: [/127.0.0.1:9999] Address already in use
> ```
> 
> Third, after restarting everything, was same as the first. I had a non-default wallet set.

Fixed in 6063a4b , forgot to replace a few variables :man_facepalming:

**Comment by user411** [comment]
Loading a wallet with a password set seemed to succeed, but next call to `walletinfo` said I have the default wallet set.
Screen Shot 2022-07-19 at 12 40 12 PM

When I told it to load the wallet with a password again, a myriad of errors pop up on the backend.
```
[info] 2022-07-19T18:43:39UTC INFO [Server] Unhandled error in server:
[info] java.util.concurrent.RejectedExecutionException: Task slick.basic.BasicBackend$DatabaseDef$$anon$3@2ad288fa rejected from slick.util.AsyncExecutor$$anon$1$$anon$2@561d1f8e[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 2771]
[info] 	at java.base/java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2055)
[info] 	at java.base/java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:825)
[info] 	at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1355)
[info] 	at slick.util.AsyncExecutor$$anon$1$$anon$4.execute(AsyncExecutor.scala:161)
[info] 	at slick.basic.BasicBackend$DatabaseDef.runSynchronousDatabaseAction(BasicBackend.scala:265)
[info] 	at slick.basic.BasicBackend$DatabaseDef.runSynchronousDatabaseAction$(BasicBackend.scala:263)
[info] 	at slick.jdbc.JdbcBackend$DatabaseDef.runSynchronousDatabaseAction(JdbcBackend.scala:37)
[info] 	at slick.basic.BasicBackend$DatabaseDef.slick$basic$BasicBackend$DatabaseDef$$runInContextInline(BasicBackend.scala:242)
[info] 	at slick.basic.BasicBackend$DatabaseDef.runInContextSafe(BasicBackend.scala:148)
[info] 	at slick.basic.BasicBackend$DatabaseDef.runInContext(BasicBackend.scala:142)
[info] 	at slick.basic.BasicBackend$DatabaseDef.runInContext$(BasicBackend.scala:141)
[info] 	at slick.jdbc.JdbcBackend$DatabaseDef.runInContext(JdbcBackend.scala:37)
[info] 	at slick.basic.BasicBackend$DatabaseDef.runInternal(BasicBackend.scala:77)
[info] 	at slick.basic.BasicBackend$DatabaseDef.runInternal$(BasicBackend.scala:76)
[info] 	at slick.jdbc.JdbcBackend$DatabaseDef.runInternal(JdbcBackend.scala:37)
[info] 	at slick.basic.BasicBackend$DatabaseDef.run(BasicBackend.scala:74)
[info] 	at slick.basic.BasicBackend$DatabaseDef.run$(BasicBackend.scala:74)
[info] 	at slick.jdbc.JdbcBackend$DatabaseDef.run(JdbcBackend.scala:37)
[info] 	at org.bitcoins.db.SafeDatabase.$anonfun$run$1(CRUD.scala:168)
[info] 	at akka.dispatch.MonitorableThreadFactory$AkkaForkJoinWorkerThread$$anon$3.block(ThreadPoolBuilder.scala:174)
[info] 	at java.base/java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3118)
[info] 	at akka.dispatch.MonitorableThreadFactory$AkkaForkJoinWorkerThread.blockOn(ThreadPoolBuilder.scala:172)
[info] 	at akka.dispatch.BatchingExecutor$BlockableBatch.blockOn(BatchingExecutor.scala:116)
[info] 	at scala.concurrent.package$.blocking(package.scala:124)
[info] 	at org.bitcoins.db.SafeDatabase.run(CRUD.scala:168)
[info] 	at org.bitcoins.db.CRUD.count(CRUD.scala:134)
[info] 	at org.bitcoins.db.models.MasterXPubDAO.existsOneXpub(MasterXPubDAO.scala:111)
[info] 	at org.bitcoins.db.util.DBMasterXPubApi.seedExists(DBMasterXPubApi.scala:19)
[info] 	at org.bitcoins.db.util.DBMasterXPubApi.seedExists$(DBMasterXPubApi.scala:16)
[info] 	at org.bitcoins.wallet.config.WalletAppConfig.seedExists(WalletAppConfig.scala:35)
[info] 	at org.bitcoins.wallet.config.WalletAppConfig.$anonfun$start$3(WalletAppConfig.scala:189)
[info] 	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:470)
[info] 	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
[info] 	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
[info] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[info] 	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
[info] 	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
[info] 	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
[info] 	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
[info] 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
[info] 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
[info] 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
[info] 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
[info] 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
[info] 2022-07-19T18:43:39UTC INFO [Server] Unhandled error in server:
[info] org.bitcoins.wallet.WalletNotInitialized: The wallet is not initialized
...
```

**Comment by Christewart** [comment]
2d12a9632c4cf7f6f05e219b1edaef288000cdfb now (mostly) works with `loadwallet` honoring long running tasks like processing a block.

If a user calls `loadwallet` while our wallet is processing a block in the background, the `loadwallet` command will wait until the wallet is done processing the block, and then it will switch to the new wallet. This means that `loadwallet` can take awhile depending on what is going on in the background. It can take > 30 seconds for our wallet to process a block. This means `loadwallet` but need to wait a bit of time until the call will complete.

**Pull request merged**